utils/yoke/derive/src/lib.rs

```diff
@@ -0,0 +1,237 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+//! Custom derives for `Yokeable` and `ZeroCopyFrom` from the `yoke` crate.
+
+use proc_macro::TokenStream;
+use proc_macro2::{Span, TokenStream as TokenStream2};
+use quote::quote;
+use syn::spanned::Spanned;
+use syn::{parse_macro_input, DeriveInput, Ident, Lifetime, Type};
+use synstructure::Structure;
+
+/// Custom derive for `yoke::Yokeable`,
+///
+/// If this fails to compile for lifetime issues, it means that
+/// the lifetime is not covariant and `Yokeable` is not
+///
+/// Note that right now this will fail to compile on structs involving
+/// `zerovec::ZeroMap`.
+/// Please comment on https://github.com/unicode-org/icu4x/issues/844
+/// if you need this
+#[proc_macro_derive(Yokeable)]
+pub fn yokeable_derive(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+    TokenStream::from(yokeable_derive_impl(&input))
+}
+
+fn yokeable_derive_impl(input: &DeriveInput) -> TokenStream2 {
```

Comment:
Thought: I see, so you allow `#[derive(Yokeable)]` to "always" work, and then `#[derive(ZeroCopyFrom)]` is the one that fails if the struct contains non-zero-copy fields, with an escape hatch.  I think this is fine.

Reply:
Yeah, because Yokeable is always safe to implement provided that the relevant lifetime is covariant. The implementations of `transform*()` automatically enforce covariance, so we're good there